### PR TITLE
byteorder: use gcc intrinsics for byteswap

### PR DESCRIPTION
**Pull request by tchaikov**
and detect endianness using cmake

### CODE REVIEW

File: src/rbd_replay/ActionTypes.cc [hunk 1]
```diff
@@ -26,7 +26,7 @@ void decode_big_endian_string(std::string &str, bufferlist::iterator &it) {
 #if defined(CEPH_LITTLE_ENDIAN)
   uint32_t length;
   ::decode(length, it);
-  length = swab32(length);
```
Comment (tchaikov) [inline src/rbd_replay/ActionTypes.cc]: @dillaman the rbd_replay/ActionTypes.cc and rbd-nbd.cc are updated to accommodate the changes in `byteswap.h`, could you help take a look?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -824,6 +824,12 @@ if (NOT WITH_SYSTEM_ROCKSDB)
 
 endif(NOT WITH_SYSTEM_ROCKSDB)
 
+include(TestBigEndian)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: @cbodley could you take a look at the cmake changes? `acconfig.h` will be included by `byteorder.h` in 199a26270d3f3f07a747d577e86546e4f6b0fa6e after this change. so the this change should be backward compatible.

File: src/rbd_replay/ActionTypes.cc (hunk 1 above)
Comment (dillaman) [inline src/rbd_replay/ActionTypes.cc]: lgtm

File: src/CMakeLists.txt (hunk 2 above)
Comment (cbodley) [inline src/CMakeLists.txt]: thanks, looks good 👍

File: src/include/byteorder.h [hunk 3]
```diff
@@ -67,11 +82,9 @@ MAKE_LE_CLASS(32)
...
 #undef MAKE_LE_CLASS
 
-#endif /* __cplusplus */
-
-#define init_le64(x) { (__u64)mswab64(x) }
-#define init_le32(x) { (__u32)mswab32(x) }
-#define init_le16(x) { (__u16)mswab16(x) }
+#define init_le64(x) { (__u64)mswab(x) }
+#define init_le32(x) { (__u32)mswab(x) }
+#define init_le16(x) { (__u16)mswab(x) }
```
Comment (liewegas) [inline src/include/byteorder.h]: seems like this needs the template <> parameter since we don't know that x has the right type

**Comment by tchaikov** [comment]
@liewegas fixed and repushed.

File: src/libradosstriper/RadosStriperImpl.cc [hunk 4]
```diff
@@ -103,13 +103,13 @@
...
- fl_object_stripe_unit: init_le32(0),
- fl_unused: init_le32(-1),
- fl_pg_pool : init_le32(-1),
+  init_le32(1<<22),	// fl_stripe_unit
+  init_le32(1),		// fl_stripe_count
+  init_le32(1<<22),	// fl_object_size
+  init_le32(0),		// fl_cas_hash
+  init_le32(0),		// fl_object_stripe_unit
+  init_le32(-1),	// fl_unused
+  init_le32(-1),	// fl_pg_pool
```
Comment (tchaikov) [inline src/libradosstriper/RadosStriperImpl.cc]: this change resolves http://ceph-devel.vger.kernel.narkive.com/VJwY8oYe/unable-to-compile.

**Comment by tchaikov** [comment]
changelog

- rebased against master

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-05-11_07:22:44-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2017-05-12_10:39:54-rados-wip-kefu-testing---basic-mira/

**Pull request merged**